packages/jsapi-bootstrap/src/useWidget.ts

```diff
@@ -0,0 +1,93 @@
+import { dh } from '@deephaven/jsapi-types';
+import Log from '@deephaven/log';
+import { assertNotNull } from '@deephaven/utils';
+import { useEffect, useState } from 'react';
+import { useObjectFetch } from './useObjectFetch';
+
+const log = Log.module('useWidget');
+
+/**
+ * Wrapper object for a widget and error status. Both widget and error will be `null` if it is still loading.
+ */
+type WidgetWrapper<T extends dh.Widget = dh.Widget> = {
+  /** Widget object to retrieve */
+  widget: T | null;
+
+  /** Error status if there was an issue fetching the widget */
+  error: unknown | null;
+};
+
+/**
+ * Retrieve a widget for the given variable descriptor. Note that if the widget is successfully fetched, ownership of the widget is passed to the consumer and will need to close the object as well.
+ * @param descriptor Descriptor to get the widget for
+ * @returns A WidgetWrapper object that contains the widget or an error status if there was an issue fetching the widget. Will contain nulls if still loading.
+ */
+export function useWidget<T extends dh.Widget = dh.Widget>(
+  descriptor: dh.ide.VariableDescriptor
+): WidgetWrapper<T> {
+  const [wrapper, setWrapper] = useState<WidgetWrapper<T>>(() => ({
+    widget: null,
+    error: null,
+  }));
+
+  const objectFetch = useObjectFetch<T>(descriptor);
+
+  useEffect(
+    function loadWidget() {
+      log.debug('loadWidget', descriptor);
+
+      const { fetch, error } = objectFetch;
+
+      if (error != null) {
+        // We can't fetch if there's an error getting the fetcher, just return an error
+        setWrapper({ widget: null, error });
+        return;
+      }
+
+      if (fetch == null) {
+        // Still loading
+        setWrapper({ widget: null, error: null });
+        return;
+      }
+
+      // We should be able to load the widget. Load it asynchronously, and set the widget when it's done.
+      // If we get cancelled before the fetch is done, we should close the widget and its exported objects.
+      // If not though, the consumer of the widget is expected to take ownership and close the widget appropriately.
+      let isCancelled = false;
+      async function loadWidgetInternal() {
+        try {
+          assertNotNull(fetch);
+          const newWidget = await fetch();
+          if (isCancelled) {
+            log.debug2('loadWidgetInternal cancelled', descriptor, newWidget);
+            newWidget.close();
+            newWidget.exportedObjects.forEach(
```

Comment:
I this this needs to be `newWidget.exportedObjects?.forEach`

`Widget` doesn't encompass the 1st party widgets like `Table`, `Figure`, etc. Just the plugin registered widgets

Or we need to check the descriptor type and throw/error if we try to use this with `Table`, `Figure`, etc.

Reply:
Let's throw an error if the descriptor is not for a widget.

Reply:
I instead opted to check if the `exportedObjects` is on the object and close if it is, so we handle all types now.